Initialize the Submission data structure

mgoerens opened the pull request:
This is the first step towards solving #296. The code added by this commit is currently *not* put in use in the certification pipeline and is a partial duplicate of the "checkprcontent" package.

The Submission data structure can collect and run early validation checks on all information concerning a user's submission (i.e. a GitHub Pull Request). In its current state, given an api_url (link to a given PR), it can collect the list of modified files, and extract the category, organization, chart's name and version, as well as running basic checks, such as SemVer compatibility of the provided version.

Eventually, this data structure is meant to encapsulate more information concerning the user's submission, and be dumped to a json file (using JSONEncode / JSONDecoder, more on that in a later PR) to be passed around the certification pipeline jobs as an artifact. This early (and incomplete) version is intended to trigger discussion around the design choices and general direction taken to solve #296.

komish commented on scripts/src/precheck/submission.py:
```diff
@@ -0,0 +1,220 @@
+        """Query the GitHub API in order to retrieve the list of files that are added / modified by this PR"""
+        page_number = 1
+        max_page_size, page_size = 100, 100
+        headers = {
+            "Accept": "application/vnd.github.v3+json",
+            "Authorization": f'Bearer {os.environ.get("BOT_TOKEN")}',
+        }
```
This feels odd to see here. Is this a carryover from checkpr? Should this (and the code below where we define the endpoint and make the call itself) be using https://github.com/openshift-helm-charts/development/blob/main/scripts/src/tools/gitutils.py#L70-L82

mgoerens replied:
Yes, it's another carryover. I will look into using `gitutils`.

komish replied:
If it's a carry over, you can leave it as is. I don't want you to have to change a bunch of things related to copied code if it works as is and doesn't impact what you're trying to implement.

komish commented on scripts/src/precheck/submission.py:
```diff
@@ -0,0 +1,220 @@
+            checkpr.get_file_match_compiled_patterns()
+        )
+        matches_found = 0
+        none_chart_files = {}
```
NIT: Rename this (and all references).

```suggestion
        non_chart_files = {}
```

I know this is a carryover from other code. If you want to keep it the way it is for this PR to reduce the diff surface area, that's fine.

mgoerens replied:
Good catch, I didn't even realized this typo in the original code ! I will change it.

FYI I'm not happy with the current implementation, especially of the `_parse_modified_files` method, but I needed something that works so I stole the code from checkpr. I would like to revisit it before getting this merged ideally (more to come in a further commit).

My intention was to discuss the general design before spending too much time refining the implementation.

In other word, I don't think the "diff surface area" is going to be small anyway and don't mind taking the time to rewrite part of the existing script, rather than trying to use the exact same code as much as possible.

komish replied:
> FYI I'm not happy with the current implementation, especially of the _parse_modified_files method, but I needed something that works so I stole the code from checkpr. I would like to revisit it before getting this merged ideally (more to come in a further commit).

Totally understand. 100% agree. It could use a clarity rewrite, but it works so I haven't prioritized it.

komish commented on scripts/src/precheck/submission_test.py:
```diff
@@ -0,0 +1,165 @@
+import responses
+import sys
+
+sys.path.append(".")
```
Nit: I'm assuming we're doing this here so that we can reference submission in the following line, but this is a Python-ism that bugs me a bit. I'm reading this as "add the files in `.` as import names".

Normally, you would do `from precheck import submission`, and then reference the classes as `submission.Submission`, right?

Shouldn't we have access to that without manipulating the default import directories?

For what it's worth, I do recognize that we have this in other source files, which is what makes this a nit. Feel free to disregard.

komish commented on scripts/src/precheck/submission.py:
```diff
@@ -0,0 +1,220 @@
+    def __post_init__(self):
+        """Complete the initialization of the Submission object.
+
+        modified_files should be gatered and validated at the beginning of the pipeline, as part
+        of the "pre-check" job. Hence this isn't done when a Submission object is loaded from a
+        JSON file aka when modified_files is not empty.
```
```suggestion
        modified_files should be gated and validated at the beginning of the pipeline, as part
        of the "pre-check" job. Hence this isn't done when a Submission object is loaded from a
        JSON file aka when modified_files is not empty.
```

Also, I may be misreading, but I'm not understanding the context of this comment.

mgoerens replied:
Intention was to be "gathered". I will provide more information, that was clear to me at the time of writing but as a new reader (and even to me now) it's a confusing and incomplete formulation.

komish commented on scripts/src/precheck/submission.py:
```diff
@@ -0,0 +1,220 @@
+    provenance: str = None
+
+
+@dataclass
```
Because this is a dataclass, I would understand that all of the class variables defined would be settable by default (in the generated init function). When I'm running this locally, it feels like I generally only want to pass in api_url, and the rest of the fields _should_ be derived.

Do I have a bad read on the application of this class, or should the rest of these class variables not be settable?

mgoerens replied:
Good question.

TLDR: A constructor that accepts *all* attributes is needed for the JSONDecoder to function. Specifically [here](https://github.com/openshift-helm-charts/development/pull/332/files/5a55ce326e4cde2a43d0f5ad634d7604afaea6c9#diff-ded7b02fcbdf810b9bf20e8f306d88f09f94202e4b3cf8a49149b55ab05bfcb4R39).

I envision two cases where we are going to instantiate a `Submission` object:
1. In the "pre-check" job (we can revisit the name of this job BTW), we are given an api_url, and the rest is derived. At the end of this process, we dump the Submission object as JSON to an [artifact file](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts#passing-data-between-jobs-in-a-workflow).
2. In following jobs, we want to access some of the Submission information, such as the path of the report, or the category. Either we do that directly in the jobs shell (`run`) using `jq`, or we have a python script that loads the artifact file. The later can be easily achieved with the `JSONDecoder` (Note that this is needed due to the nested objects: `Chart`, `Report`, `Source` and `Tarball`).

In order to know if we're dealing with 1. or 2., we check in `__post_init__` if we were provided with a list of `modified_files`. This way we don't populate the class again from scratch and override the given input data (e.g. artifact file) - though of course it should find the exact same info.

That's what I was trying to explain in [this comment](https://github.com/openshift-helm-charts/development/pull/332/files/5a55ce326e4cde2a43d0f5ad634d7604afaea6c9#diff-6cdd6c8509751e3be6c04f9c38d4c43c16760aab0318c9ed6cbad88bc9b89300R100-R102) that you didn't understand. I don't blame you, I haven't done an amazing job at explaining :) I will rework the comment.

komish reviewed (changes requested):
Great work here. Overall design looks good, and seems effective. Couple of thoughts in-lined.

mgoerens commented on scripts/src/precheck/submission.py:
```diff
@@ -0,0 +1,220 @@
+            or (self.version and self.version != version)
+        ):
+            msg = "[ERROR] A PR must contain only one chart. Current PR includes files for multiple charts."
+            print(msg)
```
I fully agree. There are carried over from the `checkprcontent` library. There they usually look like this:
```
                msg = "[ERROR] A PR must contain only one chart. Current PR includes files for multiple charts."
                print(msg)
                gitutils.add_output("pr-content-error-message", msg)
                exit(1)
```

Eventually I hope to deprecate `checkpr` in favor of  `Submission` (though they might have to be co-living for a period of time).

What I intend to do is to catch the Exception in the script that will call this library, and handle `print` and `gitutils.add_output` there.

So it's a left-over, kind of a reminder for my future self. I will remove the print statements as I agree it's not needed (also not to remind my future self).

mgoerens commented on scripts/src/precheck/submission.py:
```diff
@@ -0,0 +1,220 @@
+
+        while page_size == max_page_size:
+            files_api_query = f"{files_api_url}?per_page={page_size}&page={page_number}"
+            print(f"[INFO] Query files : {files_api_query}")
```
I will keep this in mind, thanks

mgoerens commented:
@komish I removed the DRAFT status as I believe this is now mature enough to be reviewed again and eventually merged with the current set of capabilities. I intend to continue iterate over this work, but this PR is already long enough, and I'd rather create smaller PR in the future.

This PR has seen significant changes since you reviewed it first:
- Most impacting change is that I account for another type of "Submission", when a user submits only the OWNERS file.
- Consequently, I added the `is_valid_certification_submission` and `is_valid_owners_submission` methods.
- Reworked tests

komish reviewed (approved):
/lgtm
